### PR TITLE
fix(core): single value enums (`const`) are not generated correctly in `3.1` specs

### PR DESCRIPTION
**Pull request by joscha**
This pull request extends the `OpenApiNormalizer` to translate single const enums (3.1) into an enum with a single value (3.0).
The fix and a unit test is in 9d36e993102d7573977dd0b62284005a1de287df
The diff of the samples produced by that fix can be seen in 2760c37b8b2768648efba5eb96f9c3beb184740e

Background:

Currently [single value enums](https://www.speakeasy.com/openapi/schemas/enums#constants-for-single-value-enums) in `3.1` are not generated correctly.

In `3.1` these two schemas should produce equivalent code:

```yaml
      SingleValueEnum_3_1:
        required:
          - type      
        properties:
          type:
            const: this-is-my-only-value
            type: string
      SingleValueEnum_3_0:
        required:
          - type
        properties:
          type:
            enum:
              - this-is-my-only-value
```

(the `type` property of each model can only ever have the value `this-is-my-only-value`. For `3.1` specs this seems to be widened to `string`, which is only technically correct.
It is especially visible in the `typescript` generator, where values are now widened to `string` instead of a `"this-is-my-only-value" as const` type. //cc @macjohnny 

Real-world example (a spec that was upgraded from `3.0.1` to `3.1.0`: 

Screenshot 2024-09-26 at 10 06 41 AM

via https://github.com/planet-a-ventures/affinity-node/pull/49

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is ~targeting a particular programming language~ core: @wing328

### CODE REVIEW

File: samples/client/others/typescript/builds/enum-single-value/models/SingleValueEnum31.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
+ * https://openapi-generator.tech
+ * Do not edit the class manually.
+ */
+
+import { HttpFile } from '../http/http';
+
+export class SingleValueEnum31 {
+    'type': string;
```
Comment (joscha) [inline samples/client/others/typescript/builds/enum-single-value/models/SingleValueEnum31.ts:16]: this should be:
```suggestion
    'type': 'this-is-my-only-value';
```

**Comment by joscha** [comment]
@wing328 I can see that you committed a mustache file that has an [`isConstEnumParam`](https://github.com/search?q=repo%3AOpenAPITools%2Fopenapi-generator%20isConstEnumParam&type=code) template: https://github.com/OpenAPITools/openapi-generator/blob/645626d2f597c6a877e4845edce46bc20ad8284f/modules/openapi-generator/src/main/resources/typescript-nestjs/api.service.mustache#L100-L118

via: https://github.com/OpenAPITools/openapi-generator/pull/8522 but I can't actually see `isConstEnumParam` mentioned anywhere else in the code - is this some magic somewhere? Can you point me to it, so I can see if it works similar to what I am trying to fix? Or is it a remnant of code in the pull request and it only ever goes into the else path of that template?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/OpenAPINormalizer.java [hunk 2]
```diff
@@ -1251,6 +1251,12 @@ private Schema processNormalize31Spec(Schema schema, Set<Schema> visitedSchemas)
             schema.getTypes().remove("null");
         }
 
+        // process const
```
Comment (joscha) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/OpenAPINormalizer.java:1254]: This is the actual fix that transforms a single const into an enum with one value, so the 3.1 and 3.0 representations are equivalent

File: samples/client/others/typescript/builds/enum-single-value/models/SingleValueEnum31.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
+ * https://openapi-generator.tech
+ * Do not edit the class manually.
+ */
+
+import { HttpFile } from '../http/http';
+
+export class SingleValueEnum31 {
+    'type': SingleValueEnum31TypeEnum;
```
Comment (joscha) [inline samples/client/others/typescript/builds/enum-single-value/models/SingleValueEnum31.ts:16]: this is the fix (1)

File: samples/client/others/typescript/builds/enum-single-value/models/SingleValueEnum31.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    static getAttributeTypeMap() {
+        return SingleValueEnum31.attributeTypeMap;
+    }
+
+    public constructor() {
+    }
+}
+
+export enum SingleValueEnum31TypeEnum {
+    ThisIsMyOnlyValue = 'this-is-my-only-value'
```
Comment (joscha) [inline samples/client/others/typescript/builds/enum-single-value/models/SingleValueEnum31.ts:39]: and (2)

File: samples/client/others/typescript/builds/enum-single-value/models/SingleValueEnum31.ts (hunk 1 above)
Comment (joscha) [reply samples/client/others/typescript/builds/enum-single-value/models/SingleValueEnum31.ts]: fix in [9d36e99](https://github.com/OpenAPITools/openapi-generator/pull/19696/commits/9d36e993102d7573977dd0b62284005a1de287df)

**Review by macjohnny — APPROVED** [review]
thanks!

**Comment by wing328** [comment]
thanks for the PR

cc @OpenAPITools/generator-core-team

**Comment by joscha** [comment]
> thanks for the PR
> 
> cc @OpenAPITools/generator-core-team 

Thank you. Sorry, wasn't aware there was a core team alias, will use that one from now on.

**Comment by joscha** [comment]
Is there anything I can do to help this one along? I am somewhat blocked in https://github.com/planet-a-ventures/affinity-node/pull/49 until this issue has been resolved.

**Comment by wing328** [comment]
lgtm. thanks for the PR

**Comment by wing328** [comment]
fyi. samples updated via https://github.com/OpenAPITools/openapi-generator/commit/c8fad4232439cc3d438563196627cba4ca1ce626

**Pull request merged**